### PR TITLE
Move running related code to separate crate

### LINKED ISSUE #839: Split forge crate
`forge` crate encapsulates to many responsibilities

We should split it into `cli` and `test-runner`

### PR DESCRIPTION
**Pull request by cptartur**
Closes #839 

## Introduced changes

- Split forge into two crates. Note that this PR will not pass checks as it is incomplete. See related PRs linked in comments for continuation of this work.

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Comment by cptartur** [comment]
Continuation: 
#1032 
#1046

File: crates/forge-runner/src/fuzzer.rs [hunk 1]
```diff
@@ -1,8 +1,9 @@
-use crate::fuzzer::arguments::CairoType;
 use anyhow::Result;
 use rand::rngs::StdRng;
 use rand::Rng;
 
+use crate::fuzzer::arguments::CairoType;
+
```
Comment (Arcticae) [inline crates/forge-runner/src/fuzzer.rs:6]: plz revert this one

File: crates/forge-runner/src/lib.rs
Comment (Arcticae) [inline crates/forge-runner/src/lib.rs:1]: Sidenote: This whole file encapsulates so many different things that it's hard to grasp what's really happening here, i would strongly advise refactoring this

File: crates/forge-runner/src/test_crate_summary.rs [hunk 2]
```diff
@@ -8,28 +8,31 @@ pub struct TestCrateSummary {
     pub test_case_summaries: Vec<TestCaseSummary>,
     /// Status of the runner after executing tests in the file
     pub runner_exit_status: RunnerStatus,
-    /// Type of the test crate
-    pub test_crate_type: CrateLocation,
+    // /// Type of the test crate
+    // pub test_crate_type: CrateLocation,
```
Comment (Arcticae) [inline crates/forge-runner/src/test_crate_summary.rs:12]: Either delete or uncomment this

File: crates/forge-runner/src/test_execution_syscall_handler.rs [hunk 3]
```diff
@@ -542,9 +542,9 @@ impl TestExecutionSyscallHandler<'_, '_> {
                         write_cheatcode_panic(&mut buffer, &panic_data);
                         Ok(())
                     }
-                    CallContractResult::Failure(CallContractFailure::Error { msg }) => Err(
-                        EnhancedHintError::from(HintError::CustomHint(Box::from(msg))),
-                    ),
+                    CallContractResult::Failure(CallContractFailure::Error { msg }) => {
+                        Err(EnhancedHintError::from(CustomHint(Box::from(msg))))
+                    }
```
Comment (Arcticae) [inline crates/forge-runner/src/test_execution_syscall_handler.rs:547]:
```suggestion
                    CallContractResult::Failure(CallContractFailure::Error { msg }) => 
                        Err(EnhancedHintError::from(CustomHint(Box::from(msg))))

```

File: crates/forge-runner/src/test_execution_syscall_handler.rs
Comment (Arcticae) [inline crates/forge-runner/src/test_execution_syscall_handler.rs:1]: All those changes are unrelated

File: crates/forge-runner/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+        let error = CancellationToken::new();
+        Self { exit_first, error }
+    }
+}
+
+impl Default for CancellationTokens {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```
Comment (piotmag769) [inline crates/forge-runner/src/lib.rs:168]: To be rebased after we get rid of them xD

File: crates/forge-runner/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+        Self {
+            sierra_program,
+            test_cases,
+        }
+    }
+}
+
+pub trait TestCaseFilter {
+    fn should_be_run(&self, test_case: &TestCase) -> bool;
+}
```
Comment (piotmag769) [inline crates/forge-runner/src/lib.rs:227]: Everything from the beginning of the file to this line should be moved to separate files

File: crates/forge-runner/src/test_crate_summary.rs (hunk 2 above)
Comment (piotmag769) [reply crates/forge-runner/src/test_crate_summary.rs:12]: This should stay uncommented for sure

File: crates/forge/src/pretty_printing.rs [hunk 6]
```diff
@@ -41,57 +42,6 @@ pub(crate) fn print_test_seed(seed: u64) {
     println!("{}: {seed}", style("Fuzzer seed").bold());
 }
 
-pub(crate) fn print_test_result(test_result: &TestCaseSummary) {
```
Comment (piotmag769) [inline crates/forge/src/pretty_printing.rs:44]: What happened with that?

File: crates/forge-runner/src/test_case_summary.rs [hunk 7]
```diff
@@ -124,7 +127,7 @@ impl TestCaseSummary {
     #[must_use]
     pub(crate) fn from_run_result_and_info(
         run_result: RunResult,
-        test_case: &TestCaseRunnable,
+        test_case: &TestCase,
```
Comment (piotmag769) [inline crates/forge-runner/src/test_case_summary.rs:130]: It should stay the way it was

File: crates/forge-runner/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+        Ok(ForkConfig {
+            url: value.url.parse()?,
+            block_id: value.block_id,
+        })
+    }
+}
+
+impl ForkConfigTrait for ForkConfig {}
+
+pub type TestCase = CollectedTestCase<ForkConfig>;
```
Comment (piotmag769) [inline crates/forge-runner/src/lib.rs:207]: Randomly changed name, making it hard to review

File: crates/forge-runner/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+
+    function
+        .signature
+        .param_types
+        .iter()
+        .filter(|pt| !builtins.contains(&pt.debug_name))
+        .collect()
+}
+
+fn print_test_result(test_result: &TestCaseSummary) {
```
Comment (piotmag769) [inline crates/forge-runner/src/lib.rs:545]: Why is it here?

File: crates/forge-runner/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+use futures::StreamExt;
+use once_cell::sync::Lazy;
+use scarb_artifacts::StarknetContractArtifacts;
+use smol_str::SmolStr;
+use starknet::core::types::BlockId;
+use std::collections::HashMap;
+use std::sync::Arc;
+use test_collector::FuzzerConfig;
+use test_collector::TestCase as CollectedTestCase;
+use test_collector::{ForkConfig as ForkConfigTrait, LinkedLibrary, RawForkParams};
```
Comment (piotmag769) [inline crates/forge-runner/src/lib.rs:24]: Those aliases make it unreadable, please remove them. If you want to rename them do it in separate PR

File: crates/forge-runner/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+    /// Runner exited without problems
+    Default,
+    /// Some test failed
+    TestFailed,
+    /// Runner did not run, e.g. when test cases got skipped
+    DidNotRun,
+}
+
+#[derive(Debug, Clone)]
+pub struct ForkConfig {
```
Comment (piotmag769) [inline crates/forge-runner/src/lib.rs:182]: randomly changed name

**Review by piotmag769 — COMMENTED** [review]
Please revert the name changes, it creates a lot of mess when moving code between files

File: crates/forge-runner/src/test_crate_summary.rs (hunk 2 above)
Comment (cptartur) [reply crates/forge-runner/src/test_crate_summary.rs:12]: We no longer pass CrateLocation to forge_runner, because it's irrelevant for running tests, only for printing statistics, which is handled in a different way. Removing this.

File: crates/forge-runner/src/test_execution_syscall_handler.rs [hunk 12]
```diff
@@ -825,9 +825,7 @@ fn write_call_contract_response(
                     .map(StarknetConversions::to_stark_felt)
                     .collect(),
             },
-            CallContractFailure::Error { msg, .. } => {
-                return Err(HintError::CustomHint(Box::from(msg)))
-            }
+            CallContractFailure::Error { msg, .. } => return Err(CustomHint(Box::from(msg))),
```
Comment (cptartur) [inline crates/forge-runner/src/test_execution_syscall_handler.rs:828]: This change is introduced by `cargo fmt`

File: crates/forge-runner/src/lib.rs (hunk 10 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:24]: It's either these aliases or introducing weird names to structs defined in this file

File: crates/forge-runner/src/lib.rs (hunk 5 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:227]: In a separate PR

File: crates/forge-runner/src/lib.rs (hunk 9 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:545]: I can either copy this code to this file or copy the whole `pretty_printing` to `forge-runner` which doesn't make much sense, because we cannot have a cyclic dependency between `forge-runner` and `forge` crates.

Ideally we'd move it to a separate crate or introduce some different abstraction, but that's for future PRs.

File: crates/forge-runner/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+    #[must_use]
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    #[must_use]
+    pub fn params(&self) -> &RawForkParams {
+        &self.params
+    }
+}
```
Comment (Arcticae) [inline crates/forge-runner/src/lib.rs:70]: This impl is not really useful, can't we just mark the fields `pub` and avoid adding `new` method if it doesn't wrap any logic/add any context?

File: crates/forge-runner/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,491 @@
...
+    }
+    #[must_use]
+    pub fn corelib_path(&self) -> &Utf8PathBuf {
+        &self.corelib_path
+    }
+}
+
+impl RunnerParams {
+    #[must_use]
+    pub fn new(
```
Comment (Arcticae) [inline crates/forge-runner/src/lib.rs:130]: Same as above

File: crates/forge-runner/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+
+#[derive(Debug, Clone)]
+pub struct TestCrate {
+    sierra_program: Program,
+    test_cases: Vec<TestCaseRunnable>,
+}
+
+impl TestCrate {
+    #[must_use]
+    pub fn new(sierra_program: Program, test_cases: Vec<TestCaseRunnable>) -> Self {
```
Comment (Arcticae) [inline crates/forge-runner/src/lib.rs:194]: same as above

File: crates/forge-runner/src/lib.rs (hunk 9 above)
Comment (Arcticae) [reply crates/forge-runner/src/lib.rs:442]: It should be at least in a separate file

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: crates/forge-runner/src/lib.rs
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:1]: #1091

File: crates/forge-runner/src/lib.rs (hunk 14 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:130]: I want to keep fields of this struct and other completely private/inaccessible, but we still have to allow constructing instances of it.

File: crates/forge-runner/src/lib.rs (hunk 9 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs]: Moved

File: crates/forge/src/lib.rs [hunk 16]
```diff
@@ -298,265 +164,26 @@ pub async fn run(
...
-        let mocked_tests = CompiledTestCrate {
+        let mocked_tests = CompiledTestCrateRaw {
             sierra_program: Program {
                 type_declarations: vec![],
                 libfunc_declarations: vec![],
                 statements: vec![],
                 funcs: vec![],
             },
-            test_cases: vec![TestCase {
+            test_cases: vec![TestCase::<RawForkConfig> {
```
Comment (piotmag769) [inline crates/forge/src/lib.rs:186]: Is it needed?

File: crates/forge/src/lib.rs [hunk 17]
```diff
@@ -298,265 +164,26 @@ pub async fn run(
...
+    use forge_runner::ForkTarget;
     use starknet::core::types::BlockId;
     use starknet::core::types::BlockTag::Latest;
-    use test_collector::ExpectedTestResult;
+    use test_collector::{ExpectedTestResult, TestCase};
 
     #[test]
     fn to_runnable_unparsable_url() {
-        let mocked_tests = CompiledTestCrate {
+        let mocked_tests = CompiledTestCrateRaw {
```
Comment (piotmag769) [inline crates/forge/src/lib.rs:179]: Is it needed?

File: crates/forge/src/lib.rs [hunk 18]
```diff
@@ -243,11 +113,6 @@ pub async fn run(
     let not_filtered: usize = test_crates.iter().map(|tc| tc.test_cases.len()).sum();
     let filtered = all_tests - not_filtered;
 
-    let test_crates = test_crates
-        .into_iter()
-        .map(|ctc| to_runnable(ctc, &runner_config))
-        .collect::<Result<Vec<_>>>()?;
```
Comment (piotmag769) [inline crates/forge/src/lib.rs:249]: Unrelated + not needed imo

File: crates/forge/src/lib.rs [hunk 19]
```diff
@@ -167,32 +38,32 @@ fn replace_id_with_params(
...
+            Ok(fork_target_from_runner_config.params().clone())
         }
     }
 }
 
 fn to_runnable(
     compiled_test_crate: CompiledTestCrateRaw,
     runner_config: &RunnerConfig,
-) -> Result<CompiledTestCrateRunnable> {
+) -> Result<TestCrate> {
```
Comment (piotmag769) [inline crates/forge/src/lib.rs:54]: Why?

File: crates/forge-runner/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+        })
+    }
+}
+
+impl ForkConfig for ValidatedForkConfig {}
+
+pub type TestCaseRunnable = TestCase<ValidatedForkConfig>;
+
+#[derive(Debug, Clone)]
+pub struct TestCrate {
```
Comment (piotmag769) [inline crates/forge-runner/src/lib.rs:189]: Can we keep the previous name? It is not actually a `TestCrate` itself, it has test configs too

File: crates/forge/src/lib.rs (hunk 18 above)
Comment (cptartur) [reply crates/forge/src/lib.rs:249]: It's needed: The `TestCrate` struct used by forge-runner doesn't have the crate location unnecessary since it's unused there at all. We need to first call `print_running_tests` and only then convert to runnable.

File: crates/forge-runner/src/lib.rs (hunk 20 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:189]: It doesn't make sense to name it `CompiledTestCrateRunnable` if it's the only variant of this struct present in forge-runner. It's no longer related to struct from the test collector except for the similar contents.

File: crates/forge/src/lib.rs (hunk 19 above)
Comment (cptartur) [reply crates/forge/src/lib.rs:54]: The forge-runner now owns the type it runs and it's named like that, so it had to be changed.

File: crates/forge-runner/src/lib.rs (hunk 13 above)
Comment (Arcticae) [reply crates/forge-runner/src/lib.rs:72]: Let's discuss offline

**Review by Arcticae — APPROVED** [review]

File: crates/forge-runner/src/lib.rs (hunk 20 above)
Comment (piotmag769) [reply crates/forge-runner/src/lib.rs:189]: I disagree, when I browse the code following the reference (cmd + leftclick) I won't check every time in what package I am, it introduces additional complexity

File: crates/forge/src/lib.rs (hunk 18 above)
Comment (piotmag769) [reply crates/forge/src/lib.rs:249]: Okay, makes sense

File: crates/forge-runner/src/lib.rs (hunk 20 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:189]: We can use `forge_runner::TestCrate` syntax instead of importing it directly if that's more readable. I don't think we names in other crates should heavily influence the names in this one, that's kind of agains the whole point of "namespacing" our code with different crates.

File: crates/forge-runner/src/lib.rs (hunk 13 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:72]: #1123

File: crates/forge-runner/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,437 @@
...
+                rec.close();
+            }
+        }
+
+        results.push(result);
+    }
+
+    // Waiting for things to finish shutting down
+    drop(send_shut_down);
+    let _ = rec_shut_down.recv().await;
```
Comment (tkumor3) [inline crates/forge-runner/src/lib.rs:284]: I think this logic is not required anymore. We wait for all tasks in prev loop - I will create ticket to remove and test it

File: crates/forge-runner/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,437 @@
+use crate::fuzzer::RandomFuzzer;
```
Comment (tkumor3) [inline crates/forge-runner/src/lib.rs:1]: This file is messy, it has +400 lines and is related to many things. IMO we should split it

File: crates/forge-runner/src/lib.rs (hunk 20 above)
Comment (piotmag769) [reply crates/forge-runner/src/lib.rs:183]: I understand your point, but I think it just adds cognitive complexity for no reason. Those crates are really tightly coupled so keeping names unique across both of them makes a lot of sense imo

**Review by piotmag769 — APPROVED** [review]

File: crates/forge-runner/src/lib.rs (hunk 20 above)
Comment (MaksymilianDemitraszek) [reply crates/forge-runner/src/lib.rs:183]: Agree with @piotmag769 here, those are important information

File: crates/forge-runner/src/lib.rs (hunk 22 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:1]: #1091 

It will be done but in a separate PR

File: crates/forge-runner/src/lib.rs (hunk 20 above)
Comment (cptartur) [reply crates/forge-runner/src/lib.rs:183]: These arguments are valid in general case, though I think that we'd never name this struct `CompiledTestCrateRunnable` or whatever if we had two separate crates from the very beginning. 

This name was introduced just because we had multiple versions of a similar struct and needed to differentiate between them which is no longer the case. 

If you look at the usages of it in the code, it's pretty much constructed in one place in forge and all other usages of it are in the forge-runner, where the name doesn't matter that much.
